mods/default/furnace.lua

```diff
@@ -167,8 +175,22 @@ local function furnace_node_timer(pos, elapsed)
 					if inv:room_for_item("dst", cooked.item) then
 						inv:add_item("dst", cooked.item)
 						inv:set_stack("src", 1, aftercooked.items[1])
+						-- stop any final replacement from clogging "src"
+						local can_cook = core.get_craft_result({
+								method = "cooking", width = 1,
+								items = {aftercooked.items[1]:to_string()}})
+
+						if can_cook.time == 0 and can_cook.item:to_string() == ""
+						and aftercooked.items[1]:to_string() ~= "" then
+							inv:set_stack("src", 1, "")
+							add_item_or_drop(inv, pos, aftercooked.items[1])
+						end
 						src_time = src_time - cooked.time
```

Comment:
Slight restructuring and some more comments to hopefully make it easier to understand.

```suggestion
						inv:add_item("dst", cooked.item)

						-- stop any final replacement from clogging "src"
						local can_cook = core.get_craft_result({
								method = "cooking", width = 1,
								items = {aftercooked.items[1]:to_string()}})
						can_cook = can_cook.time ~= 0 or not can_cook.item:is_empty()

						if aftercooked.items[1]:is_empty() or can_cook then
							-- cook the final "src" item in the next cycle
							inv:set_stack("src", 1, aftercooked.items[1])
						else
							-- the final "src" item was replaced and is not cookable
							inv:set_stack("src", 1, "")
							add_item_or_drop(inv, pos, aftercooked.items[1])
						end

						src_time = src_time - cooked.time
```